les/peer.go

```diff
@@ -48,191 +48,144 @@ var (
-		}
-	}
-	p.updateCounter += size
-	return p.updateCounter > allowedUpdateBytes
-}
+	id           string    // Peer identity.
+	version      int       // Protocol version negotiated.
+	network      uint64    // Network ID being on.
+	frozen       uint32    // Flag whether the peer is frozen.
+	announceType uint64    // New block announcement type.
+	headInfo     blockInfo // Latest block information.
 
-// freezeClient temporarily puts the client in a frozen state which means all
-// unprocessed and subsequent requests are dropped. Unfreezing happens automatically
-// after a short time if the client's buffer value is at least in the slightly positive
-// region. The client is also notified about being frozen/unfrozen with a Stop/Resume
-// message.
-func (p *peer) freezeClient() {
-	if p.version < lpv3 {
-		// if Stop/Resume is not supported then just drop the peer after setting
-		// its frozen status permanently
-		atomic.StoreUint32(&p.frozen, 1)
-		p.Peer.Disconnect(p2p.DiscUselessPeer)
-		return
-	}
-	if atomic.SwapUint32(&p.frozen, 1) == 0 {
-		go func() {
-			p.SendStop()
-			time.Sleep(freezeTimeBase + time.Duration(rand.Int63n(int64(freezeTimeRandom))))
-			for {
-				bufValue, bufLimit := p.fcClient.BufferStatus()
-				if bufLimit == 0 {
-					return
-				}
-				if bufValue <= bufLimit/8 {
-					time.Sleep(freezeCheckPeriod)
-				} else {
-					atomic.StoreUint32(&p.frozen, 0)
-					p.SendResume(bufValue)
-					break
-				}
-			}
-		}()
-	}
-}
+	// Background task queue for caching peer tasks and executing in order.
+	sendQueue *execQueue
 
-// freezeServer processes Stop/Resume messages from the given server
-func (p *peer) freezeServer(frozen bool) {
-	var f uint32
-	if frozen {
-		f = 1
-	}
-	if atomic.SwapUint32(&p.frozen, f) != f && frozen {
-		p.sendQueue.clear()
-	}
+	// Flow control agreement.
+	fcParams flowcontrol.ServerParams // The config for token bucket.
+	fcCosts  requestCostTable         // The Maximum request cost table.
+
+	closeCh chan struct{}
+	lock    sync.RWMutex // Lock used to protect all thread-sensitive fields.
 }
 
 // isFrozen returns true if the client is frozen or the server has put our
 // client in frozen state
-func (p *peer) isFrozen() bool {
+func (p *peerCommons) isFrozen() bool {
 	return atomic.LoadUint32(&p.frozen) != 0
 }
 
-func (p *peer) canQueue() bool {
+// canQueue returns an indicator whether the peer can queue a operation.
+func (p *peerCommons) canQueue() bool {
 	return p.sendQueue.canQueue() && !p.isFrozen()
 }
 
-func (p *peer) queueSend(f func()) {
-	p.sendQueue.queue(f)
+// queueSend caches a peer operation in the background task queue.
+// Please ensure to check `canQueue` before call this function
+func (p *peerCommons) queueSend(f func()) bool {
+	return p.sendQueue.queue(f)
+}
+
+// mustQueueSend starts a for loop and retry the caching if failed.
+// If the stopCh is closed, then it returns.
+func (p *peerCommons) mustQueueSend(f func(), stopCh chan struct{}) {
```

Comment:
We could also add this optional blocking feature to `execQueue` but we can keep it here if you prefer, just a thought. Also, `stopCh` parameter is always `p.closeCh` so maybe we could drop this parameter here.